Signals (#2)
temporalio/edu-interacting-with-workflows-typescript: signaling-your-workflows/developing-signals.md

# Developing Signals

## Defining a Signal

> The first step to developing with Signals is to define a Signal. A Signal is defined in your code and handled in your Workflow Definition. The name of the Signal is called a _Signal type_, and it is a string. The Signal can include a list of arguments which must be serializable. Here is an example of defining a Signal using `defineSignal`:

Author: Added

```typescript
import { defineSignal } from "@temporalio/workflow";

interface JoinInput {
  userId: string;
  groupId: string;
}

export const joinSignal = defineSignal<[JoinInput]>("join");
```

In the above example, the defined Signal - `joinSignal` - can now be used to pass on information about users joining groups. The Signal type is `join`. When `joinSignal` is received by the Workflow, it will carry with it the `userId` and `groupId` data, allowing the Workflow to react accordingly (e.g., updating its state to reflect that a specific user has joined a specific group).

## Handling a Signal

> However, just defining a Signal isn't enough. We also need to let the Workflow know what happens when it receives a Signal, which is what Signal handlers are for. Signal handlers are Workflow functions that listen for Signals by the Signal name. These handlers define how a Workflow should react when it receives a specific type of Signal. In order to implement a Signal handler, you will need to use the [`setHandler`](https://typescript.temporal.io/api/namespaces/workflow/#sethandler) function provided by the TypeScript SDK API.

Author: Fixed

Here is an example of creating a Signal handler using the Signal - `joinSignal` - we defined in the above section.

```typescript
import { setHandler } from "@temporalio/workflow";

export async function joinSignalHandler() {
  const groups = new Map<string, Set<string>>();

  setHandler(joinSignal, ({ userId, groupId }: JoinInput) => {
    const group = groups.get(groupId);
    if (group) {
      group.add(userId);
    } else {
      groups.set(groupId, new Set([userId]));
    }
  });
}
```

> When the Workflow execution is running and it receives `joinSignal`, the Signal handler function above is invoked. Since the `joinSignalHandler` handler defines what should happen when `joinSignal` is received, the function updates the `groups` map to reflect the new state: either adding a user to an existing group or creating a new group with the user. Through this mechanism, the Workflow dynamically reacts to external events without interrupting its execution flow.

Author: added

> If this function is called multiple times, the last handler will overwrite any previous calls.

Author: I was referring to if the `setHandler` function is called multiple times, the last handler will overwrite any previous calls. This means that only the last provided handler function will be active for that Signal type. I have clarified!

# Developing Signals

> 

Author: Good idea! Added preface

## Defining a Signal

The first step to developing with Signals is to define a Signal. A Signal is defined in your code and handled in your Workflow Definition. The name of the Signal is called a _Signal type_, and it is a string. The Signal can include a list of arguments which must be serializable. Here is an example of defining a Signal using the `defineSignal` API from the TypeScript SDK:

```typescript
import { defineSignal } from "@temporalio/workflow";

interface JoinInput {
  userId: string;
  groupId: string;
}

export const joinSignal = defineSignal<[JoinInput]>("join");
```

In the above example, the defined Signal - `joinSignal` - can now be used to pass on information about users joining groups. The Signal type is `join`. When `joinSignal` is received by the Workflow, it will carry with it the `userId` and `groupId` data, allowing the Workflow to react accordingly (e.g., updating its state to reflect that a specific user has joined a specific group).

## Handling a Signal

> However, just defining a Signal isn't enough. We also need to let the Workflow know what happens when it receives a Signal. This is done with a Signal handler. Signal handlers are Workflow functions that listen for Signals by the Signal name. These handlers define how a Workflow should react when it receives a specific type of Signal. In order to implement a Signal handler, you will need to use the [`setHandler`](https://typescript.temporal.io/api/namespaces/workflow/#sethandler) function provided by the TypeScript SDK API.

Author: fixed

> However, just defining a Signal isn't enough. We also need to let the Workflow know what happens when it receives a Signal. This is done with a Signal handler. Signal handlers are Workflow functions that listen for Signals by the Signal name. These handlers define how a Workflow should react when it receives a specific type of Signal. In order to implement a Signal handler, you will need to use the [`setHandler`](https://typescript.temporal.io/api/namespaces/workflow/#sethandler) function provided by the TypeScript SDK API.

Author: Fixed

Here is an example of creating a Signal handler using the Signal - `joinSignal` - we defined in the above section.

```typescript
import { setHandler } from "@temporalio/workflow";

export async function joinSignalHandler() {
  const groups = new Map<string, Set<string>>();

  setHandler(joinSignal, ({ userId, groupId }: JoinInput) => {
    const group = groups.get(groupId);
    if (group) {
      group.add(userId);
    } else {
      groups.set(groupId, new Set([userId]));
    }
  });
}
```

> When the Workflow execution is running and it receives `joinSignal`, the Signal handler function above is invoked. Since the `joinSignalHandler` handler defines what should happen when `joinSignal` is received, the function updates the `groups` map to reflect the new state: either adding a user to an existing group or creating a new group with the user. Through this mechanism, the Workflow dynamically reacts to external events without interrupting its execution flow.

Author: fixed

Signals in Temporal allow Workflows to react to external events without disrupting their ongoing execution - that is, Signals are not designed to necessarily block the overall Workflow execution. The Workflow continues to execute its logic independently of when or whether a Signal will be received. However, if required, a Workflow can be designed to pause its execution until a specific Signal is received. This is achieved using `await condition()`, a function imported from the TypeScript SDK which will stop the Workflow's progress until a certain condition is fulfilled. You will see later on in the hands-on exercise.

> If this `setHandler` function is called multiple times, the last handler will overwrite any previous calls. This means that only the last provided handler function will be active for that Signal type. For example, if you first set Handler A for Signal X and then later in the Workflow set Handler B for Signal X, only Handler B will respond to Signal X.

Author: Fixed!

> Now that we have seen what a Signal is, how to define it, and how to use it to define how a Workflow should react - let's see how to send a Signal in the following section.

Author: changed

## Defining a Signal

The first step to developing with Signals is to define a Signal. A Signal is defined in your code and handled in your Workflow Definition. A given Workflow Definition can support multiple Signals. Each is uniquely identified by the Signal type, a case-sensitive string value specified by the Client when sending a Signal. The Signal can include a list of arguments which must be serializable. Here is an example of defining a Signal using the `defineSignal` API from the TypeScript SDK:

```typescript
import { defineSignal } from "@temporalio/workflow";

interface JoinInput {
  userId: string;
  groupId: string;
}

export const joinSignal = defineSignal<[JoinInput]>("join");
```

In the above example, the defined Signal - `joinSignal` - can now be used to pass on information about users joining groups. The Signal type is `join`. When `joinSignal` is received by the Workflow, it will carry with it the `userId` and `groupId` data, allowing the Workflow to react accordingly (e.g., updating its state to reflect that a specific user has joined a specific group).

## Handling a Signal

> However, just defining a Signal isn't enough. We also need to let the Workflow know what to do when it receives a Signal of that type. This is done with a Signal handler. Signal handlers are functions defined in the Workflow that listen for Signals by the Signal name. These handlers define how a Workflow should react when it receives a specific type of Signal. In order to implement a Signal handler, you will need to use the [`setHandler`](https://typescript.temporal.io/api/namespaces/workflow/#sethandler) function provided by the TypeScript SDK API.

Reviewer: ```suggestion
However, just defining a Signal isn't enough. We also need to let the Workflow know what to do when it receives a Signal of that type. This is done with a Signal handler. Signal handlers are functions defined in the Workflow that listen for incoming Signals of a given type. These handlers define how a Workflow should react when it receives a specific type of Signal. In order to implement a Signal handler, you will need to use the [`setHandler`](https://typescript.temporal.io/api/namespaces/workflow/#sethandler) function provided by the TypeScript SDK API.
```

Here is an example of creating a Signal handler using the Signal - `joinSignal` - we defined in the above section.

```typescript
import { setHandler } from "@temporalio/workflow";

export async function joinSignalHandler() {
  const groups = new Map<string, Set<string>>();

  setHandler(joinSignal, ({ userId, groupId }: JoinInput) => {
    const group = groups.get(groupId);
    if (group) {
      group.add(userId);
    } else {
      groups.set(groupId, new Set([userId]));
    }
  });
}
```

When the Workflow Execution is running and it receives `joinSignal`, the Signal handler function above is invoked. Since the `joinSignalHandler` handler defines what should happen when `joinSignal` is received, the function updates the `groups` map to reflect the new state: either adding a user to an existing group or creating a new group with the user. Through this mechanism, the Workflow dynamically reacts to external events without interrupting its execution flow.

> Signals in Temporal allow Workflows to react to external events without disrupting their ongoing execution - that is, Signals are not designed to necessarily block the overall Workflow Execution. The Workflow continues to execute its logic independently of when or whether a Signal will be received. However, if required, a Workflow can be designed to pause its execution until a specific Signal is received. This is achieved using `await condition()`, a function imported from the TypeScript SDK which will stop the Workflow's progress until a certain condition is fulfilled. You will see later on in the hands-on exercise.

Reviewer: The pattern introduced in this paragraph (i.e. "setHandler + condition" to block a Workflow execution until a given signal is received) is very important. Can we give this concept a bit more emphasis? For example, add a subtitle "Waiting for a signal to be received", split the paragraph before "However", mention this as a very common pattern, etc.

> If this `setHandler` function is called multiple times, the last handler will take precedence over any previous calls. This means that only the last provided handler function will be active for that Signal type. For example, if you first set Handler A for a given Signal, and then later in the Workflow Definition set Handler B for the same Signal, then only Handler B will be invoked upon receiving this Signal.

Reviewer: ```suggestion
There can be only one handler for any signal type; `setHandler` multiple times for the same Signal type overwrites the existing handler. For example, if you first set Handler A for a given Signal, and then later in the Workflow Definition set Handler B for the same Signal, then only Handler B will be invoked upon receiving this Signal.
```
